src/ring_theory/hahn_series.lean

```diff
@@ -0,0 +1,382 @@
+  .. hahn_series.has_add }
+
+@[simp]
+lemma single_zero_mul_eq_smul [semiring R] {r : R} {x : hahn_series Γ R} :
+  (single 0 r) * x = r • x :=
+begin
+  ext a,
+  by_cases hr : r = 0,
+  { simp [hr] },
+  simp only [hr, smul_coeff, mul_coeff, single_support_of_ne, ne.def, not_false_iff, smul_eq_mul],
+  by_cases hx : x.coeff a = 0,
+  { simp only [hx, mul_zero],
+    rw [finset.sum_congr _ (λ _ _, rfl), finset.sum_empty],
+    ext ⟨a1, a2⟩,
+    simp only [finset.not_mem_empty, not_and, set.mem_singleton_iff, not_not,
+      finset.mem_add_antidiagonal, set.mem_set_of_eq, iff_false],
+    rintro rfl rfl,
+    rw [← hx, zero_add] },
+  transitivity ∑ (ij : Γ × Γ) in {((0 : Γ),a)}, (single 0 r).coeff ij.fst * x.coeff ij.snd,
+  { apply finset.sum_congr _ (λ _ _, rfl),
+    ext ⟨a1, a2⟩,
+    simp only [set.mem_singleton_iff, prod.mk.inj_iff, finset.mem_add_antidiagonal,
+      finset.mem_singleton, set.mem_set_of_eq],
+    split,
+    { rintro ⟨rfl, rfl, h2⟩,
+      refine ⟨rfl, (zero_add _).symm⟩ },
+    { rintro ⟨rfl, rfl⟩,
+      simp [hx] } },
+  { simp }
+end
+
+lemma mul_single_zero_coeff [semiring R] {r : R} {x : hahn_series Γ R} {a : Γ} :
+  (x * (single 0 r)).coeff a = x.coeff a * r :=
+begin
+  by_cases hr : r = 0,
+  { simp [hr] },
+  simp only [hr, smul_coeff, mul_coeff, single_support_of_ne, ne.def, not_false_iff, smul_eq_mul],
+  by_cases hx : x.coeff a = 0,
+  { simp only [hx, zero_mul],
+    rw [finset.sum_congr _ (λ _ _, rfl), finset.sum_empty],
+    ext ⟨a1, a2⟩,
+    simp only [finset.not_mem_empty, not_and, set.mem_singleton_iff, not_not,
+      finset.mem_add_antidiagonal, set.mem_set_of_eq, iff_false],
+    rintro rfl h rfl,
+    rw add_zero at hx,
+    exact h hx, },
+  transitivity ∑ (ij : Γ × Γ) in {(a,(0 : Γ))}, x.coeff ij.fst * (single 0 r).coeff ij.snd,
+  { apply finset.sum_congr _ (λ _ _, rfl),
+    ext ⟨a1, a2⟩,
+    simp only [set.mem_singleton_iff, prod.mk.inj_iff, finset.mem_add_antidiagonal,
+      finset.mem_singleton, set.mem_set_of_eq],
+    split,
+    { rintro ⟨rfl, h, rfl⟩,
+      refine ⟨(add_zero _).symm, rfl⟩ },
+    { rintro ⟨rfl, rfl⟩,
+      simp [hx] } },
+  { simp }
+end
+
+theorem support_mul_subset_mul_support [semiring R] {x y : hahn_series Γ R} :
+  {b : Γ | (x * y).coeff b ≠ 0} ⊆ {a : Γ | x.coeff a ≠ 0} + {a : Γ | y.coeff a ≠ 0} :=
+begin
+  apply set.subset.trans (λ x hx, _) finset.support_add_antidiagonal_subset_add,
+  { exact x.is_wf_support },
+  { exact y.is_wf_support },
+  contrapose! hx,
+  simp only [finset.not_nonempty_iff_eq_empty, ne.def, set.mem_set_of_eq] at hx,
+  simp [hx],
+end
+
+noncomputable instance [semiring R] : semiring (hahn_series Γ R) :=
+{ zero := 0,
+  one := 1,
+  add := (+),
+  mul := (*),
+  zero_mul := λ _, by { ext, simp },
+  mul_zero := λ _, by { ext, simp },
+  one_mul := λ x, single_zero_mul_eq_smul.trans (one_smul _ _),
+  mul_one := λ x, by { ext, exact mul_single_zero_coeff.trans (mul_one _) },
+  mul_assoc := λ x y z, by { ext b,
```

Comment:
I think that I would factor out this proof into a `private lemma`, just for the sake of readability.